platforms/documentation/docs/src/docs/userguide/best-practices/bp_tasks.adoc

```diff
@@ -121,3 +121,66 @@ include::sample[dir="snippets/bestPractices/useCacheabilityAnnotations/groovy",f
 
 === Tags
 `#tasks`, `#caching`
+
+== Don't resolve Configurations before Task Execution
+
+Resolving configuration before task execution phase can lead to incorrect results and slows gradle build performance.
+
+Once configuration is resolved it can't be resolved again, and it can lose some information when it is resolved too early.
+Calling manual resolution via link:{javadocPath}/org/gradle/api/artifacts/Configuration.html#resolve()[resolve] method returns a set of Files that doesn't preserve references to the tasks that produced these files.
+This will break automatic task dependencies wiring through task input properties.
+
+=== Example
+
+==== Don't Do This
+
+====
+include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/kotlin",files="build.gradle.kts[tags=resolve-call]"]
+include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/groovy",files="build.gradle[tags=resolve-call]"]
+====
+
+<1> *Explicit resolve call*: `runtimeClasspath` configuration will be resolved as soon as task created. It will lose all task dependencies as well.
+
+==== Do This Instead
+
+Use lazy link:{javadocPath}/org/gradle/api/file/FileCollection.html[FileCollection] API instead.
+
+====
+include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/kotlin",files="build.gradle.kts[tags=good-classpath-printer]"]
+include::sample[dir="snippets/bestPractices/avoidResolvingConfigurationsManually/groovy",files="build.gradle[tags=good-classpath-printer]"]
+====
+
+<1> *Pass runtimeClasspath as input to the File Collection*: the File Collection will not resolve `runtimeClasspath` immediately, and will preserve its task dependencies.
```

Comment:
I think the preserving task dependencies is an important part of this.  In this example there are no task dependencies to preserve, so noting that it will preserve task dependencies here is odd.

Maybe this example should focus on the other negative impacts of configuration-time resolution.  Or should include a task dependency (with the extra complexity that entails) - then it could demo the resolution properly causing the task to run and improperly not doing so if resolved at configuration time.

Reply:
I've reworked this sample entirely. Now I add separate sub-project and declare dependency on it.